Mathlib/Order/Chain.lean

```diff
@@ -0,0 +1,390 @@
+#align is_chain.insert IsChain.insert
+
+theorem isChain_univ_iff : IsChain r (univ : Set α) ↔ IsTrichotomous α r := by
+  refine' ⟨fun h => ⟨fun a b => _⟩, fun h => @isChain_of_trichotomous _ _ h univ⟩
+  rw [or_left_comm, or_iff_not_imp_left]
+  exact h trivial trivial
+#align is_chain_univ_iff isChain_univ_iff
+
+theorem IsChain.image (r : α → α → Prop) (s : β → β → Prop) (f : α → β)
+    (h : ∀ x y, r x y → s (f x) (f y)) {c : Set α} (hrc : IsChain r c) : IsChain s (f '' c) :=
+  fun _ ⟨_, ha₁, ha₂⟩ _ ⟨_, hb₁, hb₂⟩ =>
+  ha₂ ▸ hb₂ ▸ fun hxy => (hrc ha₁ hb₁ <| ne_of_apply_ne f hxy).imp (h _ _) (h _ _)
+#align is_chain.image IsChain.image
+
+section Total
+
+variable [IsRefl α r]
+
+theorem IsChain.total (h : IsChain r s) (hx : x ∈ s) (hy : y ∈ s) : x ≺ y ∨ y ≺ x :=
+  (eq_or_ne x y).elim (fun e => Or.inl <| e ▸ refl _) (h hx hy)
+#align is_chain.total IsChain.total
+
+theorem IsChain.directedOn (H : IsChain r s) : DirectedOn r s := fun x hx y hy =>
+  ((H.total hx hy).elim fun h => ⟨y, hy, h, refl _⟩) fun h => ⟨x, hx, refl _, h⟩
+#align is_chain.directed_on IsChain.directedOn
+
+protected theorem IsChain.directed {f : β → α} {c : Set β} (h : IsChain (f ⁻¹'o r) c) :
+    Directed r fun x : { a : β // a ∈ c } => f x :=
+  fun ⟨a, ha⟩ ⟨b, hb⟩ =>
+    (by_cases fun hab : a = b => by
+      simp only [hab, exists_prop, and_self_iff, Subtype.exists]
+      exact ⟨b, hb, refl _⟩)
+    fun hab => ((h ha hb hab).elim fun h => ⟨⟨b, hb⟩, h, refl _⟩) fun h => ⟨⟨a, ha⟩, refl _, h⟩
+#align is_chain.directed IsChain.directed
+
+theorem IsChain.exists3 (hchain : IsChain r s) [IsTrans α r] {a b c} (mem1 : a ∈ s) (mem2 : b ∈ s)
+    (mem3 : c ∈ s) : ∃ (z : _) (_ : z ∈ s), r a z ∧ r b z ∧ r c z := by
+  rcases directedOn_iff_directed.mpr (IsChain.directed hchain) a mem1 b mem2 with ⟨z, mem4, H1, H2⟩
+  rcases directedOn_iff_directed.mpr (IsChain.directed hchain) z mem4 c mem3 with
+    ⟨z', mem5, H3, H4⟩
+  exact ⟨z', mem5, trans H1 H3, trans H2 H3, H4⟩
+#align is_chain.exists3 IsChain.exists3
+
+end Total
+
+theorem IsMaxChain.isChain (h : IsMaxChain r s) : IsChain r s :=
+  h.1
+#align is_max_chain.is_chain IsMaxChain.isChain
+
+theorem IsMaxChain.not_superChain (h : IsMaxChain r s) : ¬SuperChain r s t := fun ht =>
+  ht.2.ne <| h.2 ht.1 ht.2.1
+#align is_max_chain.not_super_chain IsMaxChain.not_superChain
+
+theorem IsMaxChain.bot_mem [LE α] [OrderBot α] (h : IsMaxChain (· ≤ ·) s) : ⊥ ∈ s :=
+  (h.2 (h.1.insert fun _ _ _ => Or.inl bot_le) <| subset_insert _ _).symm ▸ mem_insert _ _
+#align is_max_chain.bot_mem IsMaxChain.bot_mem
+
+theorem IsMaxChain.top_mem [LE α] [OrderTop α] (h : IsMaxChain (· ≤ ·) s) : ⊤ ∈ s :=
+  (h.2 (h.1.insert fun _ _ _ => Or.inr le_top) <| subset_insert _ _).symm ▸ mem_insert _ _
+#align is_max_chain.top_mem IsMaxChain.top_mem
+
+open Classical
+
+/-- Given a set `s`, if there exists a chain `t` strictly including `s`, then `succ_chain s`
+is one of these chains. Otherwise it is `s`. -/
```

Comment:
```suggestion
/-- Given a set `s`, if there exists a chain `t` strictly including `s`, then `SuccChain s`
is one of these chains. Otherwise it is `s`. -/
```